### PR TITLE
Windows test suite compatibility

### PR DESCRIPTION
**Pull request by jmikola**
This PR is a rebased version of #10 with additional cleanup.

### CODE REVIEW

File: tests/AbstractProcessTest.php [hunk 1]
```diff
@@ -299,6 +317,49 @@ public function testTerminateWithStopAndContinueSignalsUsingEventLoop()
...
 
     /**
+     * @dataProvider provideOutputSizeAndExpectedMaxDuration
+     */
+    public function testProcessWithFixedOutputSize($size, $expectedMaxDuration = 5)
+    {
+        // Note: very strange behaviour of Windows (PHP 5.5.6):
+        // on a 1000 long string, Windows succeeds.
+        // on a 10000 long string, Windows fails to output anything.
+        // On a 100000 long string, it takes a lot of time but succeeds.
```
Comment (jmikola) [inline tests/AbstractProcessTest.php]: @moufmouf: Per your explanation in #10 and the above comments, this test fails on Windows. Does the workaround you're proposing in #11 fix this?

**Comment by jmikola** [comment]
@moufmouf: I'd appreciate it if you could run these test suite changes on Windows and report back whether they succeed. Based on earlier discussion with @auroraeosrose, I was inclined to make `bypass_shell` default to `true`. I also refactored the command-line preparation to ensure all tests are preparing the command-line consistently.

File: tests/AbstractProcessTest.php [hunk 2]
```diff
@@ -68,22 +68,29 @@ public function testGetTermSignalWhenRunning($process)
...
-        $loop->addTimer(0.001, function(Timer $timer) use ($process, &$output) {
+        $loop->addTimer(0.001, function(Timer $timer) use ($process, &$output, &$error) {
             $process->start($timer->getLoop());
-            $process->stdout->on('data', function () use (&$output) {
-                $output .= func_get_arg(0);
+            $process->stdout->on('data', function ($data) use (&$output) {
+                $output .= $data;
+            });
+            $process->stderr->on('data', function ($data) use (&$error) {
+                $error .= $data;
```
Comment (jmikola) [inline tests/AbstractProcessTest.php]: @moufmouf: Why was the stderr check only added to this test case? Was there a case where it failed on Windows and you caught some output here?

File: tests/AbstractProcessTest.php (hunk 1 above)
Comment (moufmouf) [inline tests/AbstractProcessTest.php]: @jmikola: Yes, this test fails on Windows... In PHP > 5.6.3, it fails. In PHP < 5.6.3, it locks and the test never halts, which is even worst. #11 fixes this test case.

**Comment by moufmouf** [comment]
@jmikola Here is the result of the test suite running on PHP 5.6.6 (it's a bit long, I put the raw output so you can see what is happening):

```
PHPUnit 4.5.0 by Sebastian Bergmann and contributors.

Configuration read from D:\Projects\harmony\child-process-pr-win-test-compat\php
unit.xml.dist

.S.SSSSSSSSSSSSSS.S.SSSSSSSSSSSSSS.E.ESS..ESESS.FF

Time: 33.21 seconds, Memory: 5.00Mb

There were 4 errors:

1) React\Tests\ChildProcess\StreamSelectLoopProcessTest::testSetEnhanceSigchildC
ompatibilityCannotBeCalledIfProcessIsRunning
proc_open(): CreateProcess failed, error code - 2

D:\Projects\harmony\child-process-pr-win-test-compat\src\Process.php:98
D:\Projects\harmony\child-process-pr-win-test-compat\tests\AbstractProcessTest.p
hp:31

2) React\Tests\ChildProcess\StreamSelectLoopProcessTest::testIsRunning
proc_open(): CreateProcess failed, error code - 2

D:\Projects\harmony\child-process-pr-win-test-compat\src\Process.php:98
D:\Projects\harmony\child-process-pr-win-test-compat\tests\AbstractProcessTest.p
hp:47

3) React\Tests\ChildProcess\StreamSelectLoopProcessTest::testStartAndAllowProces
sToExitSuccessfullyUsingEventLoop
proc_open(): CreateProcess failed, error code - 2

D:\Projects\harmony\child-process-pr-win-test-compat\src\Process.php:98
D:\Projects\harmony\child-process-pr-win-test-compat\tests\AbstractProcessTest.p
hp:173
D:\Projects\harmony\child-process-pr-win-test-compat\vendor\react\event-loop\Tim
er\Timers.php:90
D:\Projects\harmony\child-process-pr-win-test-compat\vendor\react\event-loop\Str
eamSelectLoop.php:177
D:\Projects\harmony\child-process-pr-win-test-compat\tests\AbstractProcessTest.p
hp:176

4) React\Tests\ChildProcess\StreamSelectLoopProcessTest::testStartAlreadyRunning
Process
proc_open(): CreateProcess failed, error code - 2

D:\Projects\harmony\child-process-pr-win-test-compat\src\Process.php:98
D:\Projects\harmony\child-process-pr-win-test-compat\tests\AbstractProcessTest.p
hp:222

--

There were 2 failures:

1) React\Tests\ChildProcess\StreamSelectLoopProcessTest::testProcessWithFixedOut
putSize with data set #1 (10000, 5)
Failed asserting that 4096 matches expected 10002.

D:\Projects\harmony\child-process-pr-win-test-compat\tests\AbstractProcessTest.p
hp:348

2) React\Tests\ChildProcess\StreamSelectLoopProcessTest::testProcessWithFixedOut
putSize with data set #2 (100000, 5)
Process took longer than expected.
Failed asserting that 32.302847146987915 is equal to 5 or is less than 5.

D:\Projects\harmony\child-process-pr-win-test-compat\tests\AbstractProcessTest.p
hp:350

FAILURES!
Tests: 45, Assertions: 34, Failures: 2, Errors: 4, Skipped: 35.
```

Note: on PHP 5.5.12, the unit test simply locks due to a PHP bug (that was fixed in PHP 5.5.18 and PHP 5.6.3)

Here is the output (until it locks and stops responding):

```
Configuration read from D:\Projects\harmony\child-process-pr-win-test-compat\php
unit.xml.dist

.S.SSSSSSSSSSSSSS.S.SSSSSSSSSSSSSS.E.ESS..ESESS..
```

**Comment by jmikola** [comment]
@moufmouf: If I understand correctly, the `testProcessWithFixedOutputSize` cases are expected to fail. Were the earlier four errors expected? I didn't see them mentioned in your previous PR.

`CreateProcess failed, error code - 2` seems terribly vague, but it appears to be due to a missing file path (see: https://github.com/mikehaertl/phpwkhtmltopdf/issues/8#issuecomment-12534300). I assume this is because those tests refer to the `sleep` and `echo` commands. Is it possible that enabling the `bypass_shell` option for `proc_open()` broke that and they were previously emulated by `cmd.exe`?

**Comment by clue** [comment]
Thank you for digging into this and filing this PR! :+1:

I still think that getting this feature in makes perfect sense and I'm looking forward to a future update here as per https://github.com/reactphp/child-process/issues/9#issuecomment-357438306!

Given that this is currently a WIP and hasn't seen any updates in a while and I don't see it's likely this will get traction any time soon, I will have to close this for now. Please don't see this as discouraging, we would still love to get this feature in and very much appreciate your effort! You can simply complete this feature and update this PR and we're happy to reopen this :+1: 

Thank you for your effort!

**Pull request closed without merging**